File: python/31. Next Permutation.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        i = len(nums) - 1
        while i >= 0 and nums[i] <= nums[i-1]:
            i -= 1
        nums[i:] = reversed(nums[i:])
        if i > 0:
            k = i 
            pivot = nums[i-1]
            while k < len(nums) and nums[k] <= pivot:
                k += 1
            nums[k], nums[i-1] = nums[i-1], nums[k]

File: cpp/1-50/31. Next Permutation.cpp
class Solution {
public:
	void nextPermutation(vector<int>& nums) {
		if (nums.empty())
			return;
		int i = nums.size() - 2;
		for (; i >= 0; i--) {
			if (nums[i] < nums[i + 1])
				break;
		}
		if (i == -1) {
			sort(nums.begin(), nums.end());
			return;
		}
		int j = nums.size() - 1;
		for(;j >= i; j--)
		{
			if(nums[j] > nums[i])
				break;
		}
		swap(nums[i], nums[j]);
		reverse(nums.begin() + i + 1, nums.end());
	}
};